src/libraries/System.Text.Json/source_generation/src/JsonSourceGenerator.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace System.Text.Json.SourceGeneration
+{
+    [Generator]
+    public sealed class JsonSourceGenerator : ISourceGenerator
+    {
+        public Dictionary<string, Type>? SerializableTypes { get; private set; }
+
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new JsonSerializableSyntaxReceiver());
+        }
+
+        public void Execute(GeneratorExecutionContext executionContext)
+        {
+            JsonSerializableSyntaxReceiver receiver = (JsonSerializableSyntaxReceiver)executionContext.SyntaxReceiver;
+            MetadataLoadContext metadataLoadContext = new(executionContext.Compilation);
+
+            TypeExtensions.NullableOfTType = metadataLoadContext.Resolve(typeof(Nullable<>));
+
+            // Discover serializable types indicated by JsonSerializableAttribute.
+            foreach (CompilationUnitSyntax compilationUnit in receiver.CompilationUnits)
+            {
+                SemanticModel compilationSemanticModel = executionContext.Compilation.GetSemanticModel(compilationUnit.SyntaxTree);
+
+                foreach (AttributeListSyntax attributeListSyntax in compilationUnit.AttributeLists)
+                {
+                    AttributeSyntax attributeSyntax = attributeListSyntax.Attributes.Single();
+                    IMethodSymbol attributeSymbol = compilationSemanticModel.GetSymbolInfo(attributeSyntax).Symbol as IMethodSymbol;
+
+                    if (attributeSymbol?.ToString().StartsWith("System.Text.Json.Serialization.JsonSerializableAttribute") == true)
```

Comment:
This is different behavior than how the Logging Generator works. In #51064, it uses `_compilation.GetTypeByMetadataName("Microsoft.Extensions.Logging.LoggerMessageAttribute")` to recognize the attribute. 

Our source generators should use the same technique so we don't have bugs in one that don't repro in the other.

Reply:
I've refactored the detection to use the same technique.